src/NewButton/button.tsx

```diff
@@ -0,0 +1,220 @@
+        box-shadow: ${get('shadows.btn.danger.hoverShadow')({theme})};
+      }
+      // focus must come before :active so that the active box shadow overrides
+      &:focus:not([disabled]) {
+        border-color: ${get('colors.btn.danger.focusBorder')({theme})};
+        box-shadow: ${get('shadows.btn.danger.focusShadow')({theme})};
+      }
+
+      &:active:not([disabled]) {
+        color: ${get('colors.btn.danger.selectedText')({theme})};
+        background-color: ${get('colors.btn.danger.selectedBg')({theme})};
+        box-shadow: ${get('shadows.btn.danger.selectedShadow')({theme})};
+        border-color: ${get('colors.btn.danger.selectedBorder')({theme})};
+      }
+
+      &:disabled {
+        color: ${get('colors.btn.danger.disabledText')({theme})};
+        background-color: ${get('colors.btn.danger.disabledBg')({theme})};
+        border-color: ${get('colors.btn.danger.disabledBorder')({theme})};
+      }
+    `,
+    invisible: `
+      color: ${get('colors.accent.fg')({theme})};
+      background-color: transparent;
+      border: 0;
+      border-radius: ${get('radii.2')({theme})};
+      box-shadow: none;
+    
+      &:disabled {
+        color: ${get('colors.primer.fg.disabled')({theme})};
+      }
+      &:focus:not([disabled]) {
+        box-shadow: ${get('shadows.btn.focusShadow')({theme})};
+      }
+      &:hover:not([disabled]) {
+        background-color: ${get('colors.btn.hoverBg')({theme})};
+      }
+      &:active:not([disabled]) {
+        background-color: ${get('colors.btn.selectedBg')({theme})};
+      }
+    `
+  }
+  return style[variant]
+}
+
+const getSizes = (size = 'medium', variant: VariantType = 'default', iconOnly: boolean) => {
+  let paddingTop, paddingLeft
+  switch (size) {
+    case 'small':
+      paddingTop = 3
+      paddingLeft = 12
+      break
+    case 'large':
+      paddingTop = 9
+      paddingLeft = 20
+      break
+    case 'medium':
+    default:
+      paddingTop = 5
+      paddingLeft = 16
+  }
+  if (iconOnly) {
+    paddingLeft = paddingTop + 2
+  }
+  if (variant === 'invisible') {
+    paddingTop = paddingTop + 1
+  }
+  return `
+    padding:${paddingTop}px ${paddingLeft}px;
+  `
+}
+const ButtonBase = styled.button<ButtonBaseProps & {iconOnly: boolean}>`
+  ${buttonBaseStyles}
+  ${props => getVariantStyles(props.theme, props.variant)}
+  ${props => getSizes(props.size, props.variant, props.iconOnly)}
+  ${sizes}
+  ${sx}
+  ${fontSize}
+`
+
+const Button = forwardRef<HTMLAnchorElement | HTMLButtonElement, ComponentProps<typeof ButtonBase>>(
+  ({children, ...props}, forwardedRef) => {
+    const {icon: Icon, caret, size} = props
+    let iconOnly = false
+    if (Icon && !children) {
+      iconOnly = true
+    }
+    const iconWrapStyles = {
+      display: 'inline-block',
+      ...(!iconOnly ? {mr: 2} : {})
+    }
+    return (
+      <ButtonBase ref={forwardedRef} {...props} iconOnly={iconOnly}>
+        {Icon && (
```

Comment:
I have some thoughts on icon only buttons 😄 firstly, I love that you pointed out they should be square in your loom 🎉 

For a11y, this is my [favorite article](https://www.sarasoueidan.com/blog/accessible-icon-buttons/) on icon only buttons from Sara Soueidan. I have personally implemented `Technique #2: Accessible Visually Hidden Text with hidden and aria-labelledby` a few times and its always worked out nicely. The one caviet is that you need to wrap `children` with a span or div, which adds another layer of dom elements (there is something quite nice about a button with just text as a child)

As a quick example:

```
<ButtonBase aria-labelledby="button-label">
{Icon && (
// icons should always be hidden as they are decorative
// aria-hidden="true" focusable="false"
   <Box sx={iconWrapStyles}>
            <Icon size={size} aria-hidden="true" focusable="false" />
    </Box>
)}
        <span id="button-label" hidden={iconOnly}>{children}</span>
</ButtonBase>
```

With this implementation, `children` is _always_ required. If `iconOnly` is `true` the label will be visually hidden and `aria-labelledby` will go to work for a screen reader. The nice thing about this API is that the consumer doesn't need to do anything other than always provide a descriptive button label and a prop `iconOnly`. 

This could also be used for a loader button if we want to announce a loading message to the user.

For buttons that have icons and a visible label, these same props should be true for icons: `aria-hidden="true" focusable="false"`

Reply:
@pksjce I was thinking more on this after our pairing session! I love how you're thinking about the public API and working towards keeping it simple. I can see how `iconOnly` bool is a bit clunky.. but I think in all Button cases, there will always need to be some kind of check like this. We'll just have to decide which one makes the most sense for the end user! For example...

`iconOnly`:
`label`: button description text, required in all variants
`leadingVisual`: pass in an icon (and maybe this should be `leadingIcon` since we don't have a use case for other types of visuals in buttons)
`iconOnly`: consumer passes `iconOnly` prop, we hide `label` programmatically

`ariaLabel`:
`ariaLabel`: button description text
`leadingVisual`: pass in an icon
In this example, instead of always requiring one label and hiding it programmatically, we're asking the consumer to provide an aria-label if its an icon button. But how could we require that without knowing its iconOnly? Further, we shouldn't rely on aria-labels if there are more solid methods available for labeling. We could ask for `ariaLabel` and then treat it exactly like `label` above with visually hidden, but that seems like an extra step for little benefit (abstracting for what reason?)

The `iconOnly` prop starts to feel like `<IconButton>` doesn't it? Which is exactly what you're addressing in this PR 😅 what if instead of `iconOnly` we call the prop `iconButton` so it _feels_ like a different component almost, but is still a prop
`<Button iconButton icon={}` maybe introduce a simple `icon` prop that if used instead of `leadingIcon` or `trailingIcon` indicates that its an iconButton?? Either way, always a prop.. always a decision placed on the consumer.

Yay buttons!

Reply:
@langermank - I was thinking along the same lines!
The three options you gave are what I also found as I hunted around for what existing design systems do.
Here's a small decision matrix for this.

No | Solution | Pros | Cons
---|---|---|---
1 | hasIconOnly prop | Clear indication of icon only | no indication of if an icon is actually given, no indication of if the button has text description
2 | ariaLabel/allyTitle | Makes sure we always have a title | We only really need the prop if an icon only button is being rendered. useless in other cases
3 | icon/leadingIcon/trailingIcon props | rest of the Button props remain same including text description, `icon` is a nice indicator of it being iconOnly and still get hold of the actual icon component | three icon related props, what happens when all three are used

I love solution 3. It makes sure we are not asking for unnecessary information from the user. It serves as a good standard. It helps our mission of the button always having description. Its a good enough indication for us to internally style the button as a square and add any extra ally properties. I will refactor the implementation to reflect this.

Reply:
I think you nailed it! Solution 3 sounds great to me. For prop collisions, I wonder if we can safeguard against that with typescript somehow.. like `if {icon} remove {leadingIcon} & {trailingIcon} from props` or maybe utilize linting with helpful messaging "did you mean `leadingIcon`?"

Reply:
🤔 I'm not sure if typescript can do something like that TBH. Technically they'd be able to use all of those props, but I might lean on making it look broken if they do it that way. linting may be possible. I will look at what I can come up with!